src/panels/BasePanel.ts

```diff
@@ -86,36 +87,29 @@ export abstract class BasePanel<TContent extends ContentId> {
     }
 }
 
-/**
- * A `MessageContext` that represents the VSCode end of the communication,
- * i.e. it posts messages to the webview, and listens to messages to VS Code.
- */
-class MessageContext<TContent extends ContentId> implements VsCodeMessageContext<TContent> {
-    constructor(
-        private readonly _webview: Webview,
-        private readonly _disposables: Disposable[]
-    ) { }
-
-    postMessage(message: ToWebviewMessage<TContent>) {
-        this._webview.postMessage(message);
-    }
-
-    subscribeToMessages(handler: ToVsCodeMessageHandler<TContent>) {
-        this._webview.onDidReceiveMessage(
-            (message: any) => {
-                if (!isValidMessage<ToVsCodeMessage<TContent>>(message)) {
-                    throw new Error(`Invalid message to VsCode: ${JSON.stringify(message)}`);
-                }
-
-                const action = (handler as MessageHandler<MessageDefinition>)[message.command];
-                if (action) {
-                    action(message.parameters, message.command);
-                } else {
-                    window.showErrorMessage(`No handler found for command ${message.command}`);
-                }
-            },
-            undefined,
-            this._disposables
-        );
-    }
-}
+function getMessageContext<TContent extends ContentId>(webview: Webview, webviewCommandKeys: CommandKeys<ToWebviewMsgDef<TContent>>, disposables: Disposable[]): VsCodeMessageContext<TContent> {
```

Comment:
We are no longer defining a class or implementing an interface here because the `MessageContext` types have a type-dependent number of members (`postXxx(...)` for each command).

Reply:
Sounds good, can you resolve the discussions and we can merge this. Thank you, WDYT?